src/mongo/jobs/DiscoverOutdatedComposites.class.php

```diff
@@ -0,0 +1,318 @@
+        }
+    }
+
+    /**
+    * Converts a composite document into an ID for finding its root CBD.
+    *
+    * @param array $compositeDocument - the composite being examined
+    * @return array - the ID of $compositeDocument's root CBD.
+    **/
+    protected function composite2cbdId($compositeDocument) {
+        // these must be valid - they come from a valid composite
+        $cbdResourceAlias = $compositeDocument[_ID_KEY][_ID_RESOURCE];
+        $cbdContextAlias = $compositeDocument[_ID_KEY][_ID_CONTEXT];
+        return array(_ID_KEY => 
+            array(_ID_RESOURCE => $cbdResourceAlias, _ID_CONTEXT => $cbdContextAlias));
+    }
+
+    /**
+    * Given an array of regeneration tasks, runs all those tasks.
+    *
+    * @param array $regenTasks
+    **/
+    public function runRegenerationTasks($regenTasks) {
+        foreach ($regenTasks as $regenTask) {
+            $specification = $regenTask->specification;
+            $compositeCollection = $regenTask->compositeCollection;
+            $compositeRegenFunction = $regenTask->compositeRegenFunction;
+            
+            // regenerate individual composites from root CBDs and specifications
+            foreach($regenTask->cbdDocuments as $cbdDoc) {
+                $compositeRegenFunction($specification, $compositeCollection, $cbdDoc);
+            }
+        }
+    }
+
+}
+
+/**
+* Represents data about a particular kind of composite, and provides tools for regenerating documents.
+**/
+class CompositeMetadata {
+    public $compositeType;
+    public $specification;
+    public $compositeCollection;
+    public $cbdCollection;
+    public $compositeRegenFunction;
+
+    public function __construct($compositeType, $specification, $compositeCollection, $cbdCollection, $compositeRegenFunction)
+    {
+        $this->compositeType = $compositeType;
+        $this->specification = $specification;
+        $this->compositeCollection = $compositeCollection;
+        $this->cbdCollection = $cbdCollection;
+        $this->compositeRegenFunction = $compositeRegenFunction;
+    }
+
+    /**
+    * Create a query component for finding composites matching matching the specification,
+    * but which are outdated.
+    *
+    * The mongodb equivalent is:
+    *
+    * { "_spec.type": <THIS TYPE>, "_spec.revision": { "$lt": <SPEC REVISION NUMBER> }}
+    *
+    * ALTERNATIVE: This is predicated on having independent revisions for each composite spec. However,
+    * It would also be possible to run this query using a single global revision and a composite spec hash:
+    *
+    * { 
+    *   "_spec.type": <THIS TYPE>, 
+    *   "_spec.revision": { "$lt": <GLOBAL REVISION NUMBER> }, 
+    *   "_spec.hash": { "$ne": <HASH OF THE SPEC> }
+    * }
+    *
+    * That alternative would make managing the revision numbers much more straightforward, since we can simply
+    * Increment that global number on every build, and the hash will tell us if an update is required. Crucially,
```

Comment:
The above method is doing most of the work for our database lookups, and deserves special scrutiny.

The method constructs a query component, which can be used to filter to outdated documents in the metadata's composite collection.